Added Support for TSQL's ISC-column_privileges

oscarno22 opened the pull request:
### Description

Currently, Babelfish does not support TSQL's
information_schema_tsql.column_privileges. This commit implements this
view and provides testing. It should be noted that the function
sys.schema_name was moved from import_export_compatibility.sql to
sys_functions.sql. The function was being called in the column_privileges view
before it was defined in import_export_compatibility.sql. This update
readjusts for the order of execution.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>
 
### Issues Resolved

- The view information_schema_tsql.column_privileges not supported.
- The function sys.schema_name being called before definition.


### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

oscarno22 committed:
Added Support for ISC-Column_Privileges

Currently, Babelfish does not support TSQL's
information_schema_tsql.column_privileges. This commit implements this
view and provides adequate testing. It should be noted that the function
sys.schema_name was moved from import_export_compatibility.sql to
sys_functions.sql. The function was being called in the column_privileges view
before it was defined in import_export_compatibility.sql. This update
readjusts for the order of execution.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Added Support for ISC-Column_Privileges

Currently, Babelfish does not support TSQL's
information_schema_tsql.column_privileges. This commit implements this
view and provides adequate testing. It should be noted that the function
sys.schema_name was moved from import_export_compatibility.sql to
sys_functions.sql. The function was being called in the column_privileges view
before it was defined in import_export_compatibility.sql. This update
readjusts for the order of execution.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated Call to column_privileges in Test

This commit updates the view call to include "where table_name is not
like 'sys%". This limits excessive output in the testing files.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Updated Call to column_privileges in Test

This commit updates the view call to include "where table_name is not
like 'sys%". This limits excessive output in the testing files.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

xhfanhe committed:
Role with members should not be allowed to be dropped (#461)

Previously user can drop a role even if the role has members. With this
change, that behavior is disallowed.

Task: BABEL-3231
Signed-off-by: Xiaohui Fanhe <hexiaohu@amazon.com>

xhfanhe committed:
Fix IS_ROLEMEMBER() and IS_MEMBER() function output (#476)

With this change,
1. IS_ROLEMEMBER() and IS_MEMBER() will return 1 when input role is PUBLIC.
2. IS_ROLEMEMBER() and IS_MEMBER() will be case-insensitive.

Task: BABEL-3228/3229
Signed-off-by: Xiaohui Fanhe <hexiaohu@amazon.com>

KushaalShroff committed:
SCOPE_IDENTITY() function and @@IDENTITY should be declared as STABLE in PG

Added Stable modifier to SCOPE_IDENTITY(), sys.babelfish_get_last_identity_numeric() and sys.babelfish_get_last_identity() functions. As far as @@IDENTITY is concerned, it internally calls sys.babelfish_get_last_identity_numeric, so even @@IDENTITY is made STABLE with this commit.

Task: BABEL-3376
Authored-by: Kushaal Shroff kushaal@amazon.com
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

shalinilohia50 committed:
In single-db, the schema resolution is incorrect when the object inside the procedure is schema qualified

The dml statement should check for the object in the schema specified
with the object, not the one specified in the procedure call.

Task: BABEL-3403
Authored-by: Shalini Lohia lshalini@amazon.com
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

skkyal committed:
For sp_describe_first_result_set changed datatype from varchar(384) to sys.nvarchar(8000)

sp_describe_first_result_set limits the T-SQL expression to 384 characters which blocks the DMS Babelfish endpoint with wide tables.
To solve this deprecated the old function/procedure and added new function/procedure by changing the datatype from varchar(384) to sys.nvarchar(8000).

Task: BABEL-3387
Authored-by:: Shlok Kumar Kyal (skkyal@amazon.com)
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

aadityavermaa committed:
Made name convention consistent in upgrade script

In the upgrade script for babelfish_extensions from 2.1.0 to 2.2.0,
there were some inconsistencies in the naming convention for deprecated
objects.

In this commit, made the naming convention consistent such that now the
deprecated objects have a suffix "deprecated_in_2_2_0".

Authored-by: Aditya Verma <adiityav@amazon.com>
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

shalinilohia50 committed:
Incorrect schema resolution for the SQL objects inside a function

When the function name is schema qualified and the SQL objects being referenced inside that function is not schema qualified, currently those objects were resolved to the default schema of the user. But that's incorrect.
With this change, the SQL objects being referenced inside the function should resolve to the function's schema.

Steps we follow:
* Search for the object in the same schema as the SQL object in which the reference occurs.
* If not found, search in the dbo schema.

This issue is specific for name resolution inside a SQL object. Babelfish resolves the schema incorrectly. The consequence of this bug is that we may return wrong results or update the wrong table.

Task: BABEL-2992
Authored-by: Shalini Lohia lshalini@amazon.com
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

timchang514 committed:
Support SET PARSEONLY ON/OFF

Adding support for PARSEONLY. We replace the plan with an empty one if PARSEONLY
is enabled. To be able to turn PARSEONLY off, we actually alter the setting
at parse-time, which appears to be the behavior in T-SQL as well.

Task: BABEL-3317
Authored-by: Tim Chang changti@amazon.com
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

surendravishnoi commented on test/JDBC/input/isc-column_privileges-prepare.txt:
```diff
@@ -0,0 +1,18 @@
+drop database if exists prep;
```
we should test in master database if possible instead of custom db. Also, use test filename as prefix to avoid duplicate names for objects.

basasairohan committed:
Converted to MVU scripts for existing system functions

Added prepare-verify scripts for some of the system functions.

Authored-by: Sai Rohan Basa <bsrohan@amazon.com>
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

shalinilohia50 committed:
In case of cross-db procedure call, incorrect schema resolution when the object inside the procedure is schema qualified

When there is cross-db procedure call and the SQL Objects referenced
inside the procedure is schema qualified, the schema resolution takes
place in a way that the schema is searched in the current database but
not the one mentioned in the cross db procedure call which is incorrect.

Suppose the cross-db procedure call is
"master.sch1.p1" and it refers to an object "sch2.t1", it should resolve
to "master_sch2.t1" not "db1_sc2.t1" which is implemented with this
change. Assuming that db1 is the current database.

Task: BABEL-3403
Authored-by: Shalini Lohia lshalini@amazon.com
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

oscarno22 committed:
Updated Test Files

This commit updates the test files to more concise names. The file names
are used as prefixes for the objects created and used within them, to
avoid duplicate objects.

TASK:

Signed-off-by: Ozzy Nolen <oscarno@amazon.com>

oscarno22 committed:
Merge branch 'oscarno-test1' of https://github.com/oscarno22/babelfish_extensions into oscarno-test1
